### ISSUE TITLE
LM v4.0 To-do List

### ISSUE THREAD

**Issue by lokka30** [ArcanePlugins/LevelledMobs#345]
###### This list is frequently edited. When a task is completed, it is deleted from the list. Tasks are regularly added; the list is not definitive at any point in time.

<details>
<summary><h1>Q/A:</h1></summary>

## What is LM4 compared to LM3?

Vastly boiling down all of our efforts here: it's a more powerful, performant, extensible, and overall refined version of LM3. It's re-programmed from the ground up. The entire plugin and its sub-systems have undergone *several* reiterations. The code has been structured in (our hopes) a future-proof fashion, and we have a decent [roadmap](https://github.com/lokka30/LevelledMobs/projects) for future LM 4.x versions.

## What tasks are available for public contribution?

Please let us know if you want to contribute! LM4 has been in development for long enough that there are only few relatively easy tasks for contributors to do, the rest will be completed solely by the maintainers (@lokka30 @stumper66 @UltimaOath). We will let you know of any tasks (other than testing) which you can do if you like. Just reach out!

Note: we need as many users as possible to test LM v4 before we release it to the public – this requires no special skills and only ~20 minutes (or more, at your option) of your time, you're just ensuring that LM4's features are working as intended and don't break even when configured correctly. 

## Is this list sorted in any particular order?

No, however, some tasks must be completed before others. For instance, the Finalisations section should only be worked on once the other parts of the To-do list (except testing) are complete.

</details>

# Quick Links:

- [Development Branch `dev-4.0`](https://github.com/lokka30/LevelledMobs/tree/4.0-dev)
- [Brainstorming Branch `bs2` (in brainstorming repo)](https://github.com/ArcanePlugins/LevelledMobs-Brainstorming/tree/bs2)

# Planner Docs:

- [Levelling Strategies](https://gist.github.com/lokka30/8800334372a6f00dc5cb37f585d98a95)
- [Commands](https://gist.github.com/lokka30/1c59e30a3b84aabd32d1668dca5949c7)

***

# To-do List:

### Custom Entity System

See #317 

### Bloodthirst Levelling Strategy

See #349 

### Variable Levelling Strategy

> Previously known as Player Levelling / Player-Mod Lvl

- Design the new system (@lokka30 and @stumper66 and @UltimaOath)
  - See #322 
- Implement the system (@lokka30 and @stumper66)

### Label System

- add an API way of retrieving the **formatted** primary label (not just the formula).
- See #378 
- Implement visibility methods and durations
- Implement and test Spigot compatibility
- Add health indicators system
- Add Boss Bar label system - see #233 

### Commands

- implement new commands (@lokka30)

### Custom Drops

- see #366 
- program custom drops system (@stumper66 and @lokka30)
  - [remember chance scaling idea](https://discord.com/channels/752310043214479462/756850940694495253/990113042962448444)
- fill in player head textures JSON file (@lokka30)
  - does this thing still even exist in LM3?
- see #333
- Chunk kill count system

### Buffs

- Add `CUSTOM_` buffs in see `util.modal.impl.ModalAttributeSet` class
- Add Shield Breaker - see #379 #329 

### Generic Tasks

- make sure all actions and conditions are programmed in (@lokka30)
- add phase of the moon condition #332 
- fill in the lists of hostile, passive etc mobs. (@lokka30 or @stumper66 or @UltimaOath)
- add a random num placeholder (between 0 and 1, inclusive): `%random%` (@lokka30)
- use a case insensitive replace instead of case sensitive replace (@lokka30)
- add placeholder/context-based boolean conditions - see #375 

### Integrations
- SimplePets
- MythicMobs
- InfernalMobs
- ...more from LM3
- EliteBosses #344 
- EliteMobs
- PyroMining #363 
- ModelEngine #367 
- RTULangAPI #372 
- CombatPets #402
- DangerousCaves 2 #273 

### Finalisations

> Not ready yet

- add migration code for customdrops.yml (LM3 -> LM4 file conversion). (@lokka30)
- add decorations to all external files (@UltimaOath and @lokka30 )
- Update README.md since LM3's and LM4's are different, stuff needs to be merged (@lokka30)
- User testing. In parallel, work on documentation (@lokka30 and @UltimaOath)
- Make LevelledMobs 4 update poster/banner, like what was used on LM3 update. (@UltimaOath)
- Release 😄 
- Publish LM4 to a Maven repository
- Update LM Items to have v4 support (@stumper66)

### Testing

> Not ready yet, but it is recommended to test each feature after it has been implemented. Baby steps are the best way to make sure we don't have to test a pile of spaghetti and rather small strands at a time :)

- Spawn-Distance and Y-Axis Levelling and Basic Levelling Strategies 
- Debug System
- Make sure ExitProcess, ExitFunction, and ExitAll are working.
  - Specifically, we need to make sure the exit status from the previous run of any function/process is not carried over.
- Make sure `SetPermanentLabelAction` and `SetPacketLabelAction` works on spigot servers

**Comment by lokka30** [comment]
Superseded by multiple issues which have been added to the v4.0 milestone
